Feature: Add file for configuring include / exclude of files to template

xortive opened the pull request:
Adds support for a file which allows for templates to configure which files should be processed, either using a whitelist method (include), or a blacklist method (exclude), with include being preferred over exclude if both are present. This mirrors the behavior used by cargo for the configuration options of the same name in `Cargo.toml`, and uses some of the same code.

closes #144
closes #124
closes #48
closes #43
closes https://github.com/cloudflare/worker-emscripten-template/issues/1

xortive commented on src/include_exclude.rs:
```diff
@@ -0,0 +1,94 @@
+    }))
+}
+
+pub fn create_default_matcher(
```
I'd like to figure out how to combine this with the create_matcher method, so it returns a single iterator, but the owned closure passed into filter makes this difficult. Would appreciate some help with making it happen!

ashleygwilliams commented:
would very much like feedback from @k0pernicus and @DD5HT on this!

k0pernicus commented:
No problem, I can look at the PR this evening :)

xortive commented on src/template.rs:
```diff
@@ -119,52 +120,57 @@ pub fn substitute(
+    //returning iterators is hard :/
+    match template_config {
+        Some(template_config) => {
+            for entry in create_matcher(&template_config, project_dir)? {
```
Clippy is angry about this line and 136 on latest stable. It's a bug in clippy:  https://github.com/rust-lang/rust-clippy/issues/4133

steveklabnik commented on src/config.rs:
```diff
@@ -0,0 +1,62 @@
+        match File::open(path) {
+            Ok(mut file) => {
+                let mut contents = String::new();
+                file.read_to_string(&mut contents)?;
```
https://doc.rust-lang.org/std/fs/fn.read_to_string.html 😄

xortive replied:
TIL. thanks!

xortive commented:
needs squash before merge

xortive commented:
renamed config file name to `cargo-generate.toml` from `.gen.toml`. I am not attached to either name, but I believe the new one is a little better.

steveklabnik commented on src/config.rs:
```diff
@@ -0,0 +1,62 @@
+use serde::Deserialize;
+use std::fs::read_to_string;
```
style nit: for types, like `use serde::Deserialize;`, importing the type fully makes sense. However, for functions, the convention is to use the module, so

    use std::fs;

here, and then below, use `fs::read_to_string`.

xortive replied:
will do!

steveklabnik commented on src/include_exclude.rs:
```diff
@@ -0,0 +1,76 @@
+
+impl Matcher {
+    pub fn new(
+        template_config: Option<TemplateConfig>,
```
Given that this function is one big `if let`/`else`, I would split it into two:

* `new` that only takes `project_dir`
* `new_with_template_config` that takes both a `project_dir` and a `TemplateConfig`

xortive replied:
and leave the option matching to the client instead? makes sense

xortive commented:
implemented steve's PR feedback, rebased onto master, and added more detail to readme about how this feature interacts with template tags in filenames.
